DuckDuckGo/DuckPlayer/YouTubePlayerNavigationHandler.swift

```diff
@@ -0,0 +1,119 @@
+        request.httpMethod = "GET"
+        return request
+    }
+
+    static func makeHTMLFromTemplate() -> String {
+        guard let html = try? String(contentsOfFile: htmlTemplatePath) else {
+            assertionFailure("Should be able to load template")
+            return ""
+        }
+        return html
+    }
+    
+    private func performNavigation(_ request: URLRequest, responseHTML: String, webView: WKWebView) {
+        // iOS 14 will be soon dropped out (and it does not support simulatedRequests)
+        if #available(iOS 15.0, *) {
+            webView.loadSimulatedRequest(request, responseHTML: responseHTML)
+        }
+    }
+    
+    private func performRequest(request: URLRequest, webView: WKWebView) {
+        let html = Self.makeHTMLFromTemplate()
+        let duckPlayerRequest = Self.makeDuckPlayerRequest(from: request)
+        performNavigation(duckPlayerRequest, responseHTML: html, webView: webView)
+    }
+}
+
+extension YoutubePlayerNavigationHandler: DuckNavigationHandling {
+    
+    func handleNavigation(_ navigationAction: WKNavigationAction,
+                          webView: WKWebView,
+                          completion: @escaping (WKNavigationActionPolicy) -> Void) {
+        if let url = navigationAction.request.url, url.isDuckPlayer {
+            let html = Self.makeHTMLFromTemplate()
+            let newRequest = Self.makeDuckPlayerRequest(from: URLRequest(url: url))
+            if #available(iOS 15.0, *) {
+                webView.loadSimulatedRequest(newRequest, responseHTML: html)
+                completion(.allow)
+                return
+            }
+        }
+        completion(.cancel)
+    }
+    
+    func handleRedirect(url: URL?, webView: WKWebView) {
+        if let url = url, url.isYoutubeVideo, !url.isDuckPlayer, let (videoID, timestamp) = url.youtubeVideoParams {
+            webView.stopLoading()
+            let newURL = URL.duckPlayer(videoID, timestamp: timestamp)
+            webView.load(URLRequest(url: newURL))
+        }
+    }
+    
+    func handleRedirect(_ navigationAction: WKNavigationAction,
+                        completion: @escaping (WKNavigationActionPolicy) -> Void,
+                        webView: WKWebView) {
+        if let url = navigationAction.request.url, url.isYoutubeVideo, !url.isDuckPlayer, let (videoID, timestamp) = url.youtubeVideoParams {
+            webView.load(URLRequest(url: .duckPlayer(videoID, timestamp: timestamp)))
+            completion(.allow)
+            return
+        }
+        completion(.cancel)
+    }
+    
+    func goBack(webView: WKWebView) {
+        guard let backURL = webView.backForwardList.backItem?.url,
+                backURL.isYoutubeVideo,
+                backURL.youtubeVideoParams?.videoID == webView.url?.youtubeVideoParams?.videoID else {
+            webView.goBack()
+            return
+        }
+        webView.goBack(skippingHistoryItems: 2)
```

Comment:
Why goBack 2? Maybe add a comment with the reason why?

Reply:
We Go from:

- Youtube list (youtube.com)
- Youtube video link (youtube.com/watch?=) which Redirects to Duck Player (duck://player/xYz)

The standard goBack() would take you to the Youtube video link, and you'll end in a loop